### PR TITLE
feat: option to automatically power off radio after inactivity

### PR DESCRIPTION
**Pull request by pault4r**
An implementation of feature request #789 for the TX16S (see also opentx issue #6421). [Note - now expanded to all radios with power buttons, see below]

Radio will automatically shutdown after a configurable number of minutes of inactivity. Option is shown on the main "Radio Setup" screen with zero minutes used to disable the feature. Shutdown will not take place if telemetry is present or if the radio is connected via USB.

I like this as it is on my radio but it might need a little more thought and development if you guys have any interest in pulling this into your repo for general use. Happy to look at it further if that is the case. 

Paul

### CODE REVIEW

**Comment by Eldenroot** [comment]
Maybe add on/off and then a configurable timer?

**Comment by pault4r** [comment]
Yes I agree the user interface would be better with with an on/off switch first and then the timer shown only if the feature is switched on, rather than just using 0 for "off"

File: companion/src/firmwares/generalsettings.cpp [hunk 1]
```diff
@@ -119,6 +119,7 @@ void GeneralSettings::init()
   backlightMode = 3; // keys and sticks
   backlightDelay = 2; // 2 * 5 = 10 secs
   inactivityTimer = 10;
+  pwrOffIfInactive = 0; // disabled
```
Comment (elecpower) [inline companion/src/firmwares/generalsettings.cpp:122]: Unnecessary as GeneralSettings initialised to null

File: radio/src/gui/colorlcd/radio_setup.cpp [hunk 2]
```diff
@@ -632,6 +632,17 @@ void RadioSetupPage::build(FormWindow * window)
   }
 #endif
 
+#if defined(PWR_BUTTON_PRESS) && defined(PCBX10)
+  // Pwr Off If Inactive
+  {
+    auto line = window->newLine(&grid);
+    new StaticText(line, rect_t{}, STR_PWR_OFF_IF_INACTIVE, 0, COLOR_THEME_PRIMARY1);
+    auto edit = new NumberEdit(line, rect_t{}, 0, 250, GET_SET_DEFAULT(g_eeGeneral.pwrOffIfInactive));
```
Comment (elecpower) [inline radio/src/gui/colorlcd/radio_setup.cpp:640]: Discrepancy for max value - Companion max is 255. The value is read from radio.yml and displayed as 255 but radio edit limit is 250.

File: companion/src/generaledit/generalsetup.ui [hunk 3]
```diff
@@ -1305,7 +1305,38 @@ p, li { white-space: pre-wrap; }
...
+        <string/>
+       </property>
+       <property name="whatsThis">
+        <string>If not zero transmitter will shut down if it has been left without inputs for the specified number of minutes.</string>
+       </property>
+       <property name="suffix">
+        <string> min</string>
+       </property>
+       <property name="maximum">
+        <number>255</number>
```
Comment (elecpower) [inline companion/src/generaledit/generalsetup.ui:1335]: Differs from radio see below

File: companion/src/firmwares/generalsettings.cpp (hunk 1 above)
Comment (pault4r) [reply companion/src/firmwares/generalsettings.cpp:122]: Agreed - thanks

File: radio/src/gui/colorlcd/radio_setup.cpp (hunk 2 above)
Comment (pault4r) [reply radio/src/gui/colorlcd/radio_setup.cpp:640]: Ah yes, well spotted :-)

File: companion/src/generaledit/generalsetup.ui (hunk 3 above)
Comment (pault4r) [reply companion/src/generaledit/generalsetup.ui:1335]: Will adjust the other one to 255....

**Comment by Eldenroot** [comment]
Is ON/OFF check box added instead of minutes?

**Comment by pault4r** [comment]
Hi @Eldenroot,

I guess we would need an ON/OFF checkbox and then the number of minutes if ON.

It's not in the code now, if you are interested I can try it that way and see how it looks then. Will get back to you on that.

If using a toggle, one question would then be why the inactivity alarm does not work in the same way (i.e. also have an ON/OFF toggle) as they are very similar functions. It could also be done that way of course.

Perhaps I should rebase this now as well as there seem to be some conflicts.

So I don't know if this feature is on any road map or has sufficient interest to be included, I just know that I find it useful (and regularly use it) and it has been requested at least once in the past. But of course that might not be enough, maybe most people do manage to switch their transmitters off :-). For me the TX16S just somehow looks like it's off most of the when the LCD is off. I've left it on a few times and moved away from it, not hearing the inactivity alarm, and I'd actually rather not have to manually switch it off a lot of the time when it can just do that itself.

Paul

**Comment by pfeerick** [comment]
Interest, definitely. And will probably be something to be included 2.10. Will just want to expand it to all radios, not just the TX16S.  I wouldn't worry about rebasing it for a week or so, as I think there are some more PRs to go that will most likely cause minor conflicts (because they'll also be touching datastructs to save their own options). As far as a checkbox that controls the edit field, that makes sense, and for sure the inactivity timer could be done the same way (for consistency).

**Comment by pault4r** [comment]
Hi Peter, 

Ah OK, sounds good :-). I will update it in in a few days then and include the checkbox. Applying to all radios with a push button switch seems like an excellent idea. I only have the TX16S to test myself hence tried not to change anything else. However, it's pretty simple and should only need a few checks on other radios to get this right.

Paul

**Comment by JimB40** [comment]
I assume this is to prevent draining radio's battery (particulary LiPo that can be destroyed once below 3.0V)
So good feature would be to have this feature based on tx voltage.
In this case no telemetry present checkig then as radio will start malfunction anyway below certain TX batt voltage.

**Comment by JimB40** [comment]
That's what I thought of. Killed radio lipo lastly. Finished dev lately and forgot to switch off tx16s while volume was down not to disturb me while in dev mode :)

**Comment by pfeerick** [comment]
Yeah, for dev work, unless you're on the move, running from battery is not a good idea. Better to have a power supply and then it doesn't matter how long you leave it turned on for 😆

**Comment by pault4r** [comment]
Have rebased the code here since it was 900 odd commits behind with some conflicts. Still the same tx16s specific patch. This the padding in the yaml generation is not quite right in this version though. To be corrected shortly.

**Comment by 3djc** [comment]
Just a thought reading this, should trainer not be checked too ?

**Comment by pault4r** [comment]
@3djc, good point regarding the trainer, I will have a look at that....

**Comment by pault4r** [comment]
Latest commit expands to all radios as mentioned by @pfeerick above.

Still the simplest interface here, 0-255 mins shown. Main lack of clarity being what 0 means. Just replacing the 0 with "off" is one way to improve but many options with a GUI as always.

**Comment by raphaelcoeffic** [comment]
@pault4r sorry for being late to the party: how does this interact with the regular inactivity timer?

**Comment by pault4r** [comment]
Hi @raphaelcoeffic, it is based on the regular inactivity timer (read only - no adjustment made to the timer).

The user could have the inactivity alarm first and then automatic power off some time after that by using both features. Or have a silent power off.

There are additional checks beyond the inactivity timer but the timer is the main thing.

**Comment by 3djc** [comment]
Just a minor comment,  the comments still mention "initially implemented for TX16S", thats not longer needed, is it ?

**Comment by raphaelcoeffic** [comment]
@pault4r if you would be so kind to rebase on top of `main`, so we can proceed with more tests.

**Comment by pault4r** [comment]
@raphaelcoeffic , now rebased :-)

**Comment by pault4r** [comment]
This implementation is basic but functional. It could benefit from additional items such as messages on the screen, countdowns, audio etc. but that's all quite subjective and could get complex for multiple radios. Hence I've only put in basic working functionality to start with. I would imagine at least one other user interface feature of some kind would be added before a final release.

**Comment by pault4r** [comment]
This is the same change as before, just rebased and with the commits organized in a clearer way

**Comment by pfeerick** [comment]
This has been working well for some time now on TX16S, T12MAX and X9D+2019. 

Some future improvements could be (mostly UI):
- probably only for colorlcd - use a toggle switch to turn on, and then display textedit field for value
- show a full screen alert when one minute remains before power off, with a new voice prompt, and haptic. Which has a countdown timer. I think it should clear itself as soon as the system is no longer inactive, thus at most perhaps a cancel button, linked to RTN, could be shown, as in the process of hitting an accept/ok button, it would clear it self :hand_over_mouth: 
- discussion points could be on whether the warning prompt should be configurable - both as to whether it appears or not, and when it should appear
- once there is an advance warning prompt, I could see room for also adding an optional low battery voltage poweroff trigger also, also with toggle switch and number edit field - for specifying the shutdown voltage - which would probably have it's max warning voltage capped to below that of the low battery alarm value.

**Pull request merged**